File: apps/api/src/app/repository/contributors.ts
import { Repository, RepositoryContributors } from '@lib/core';
import { injectable } from 'tsyringe';
import { CacheStorage } from '../service/cache-storage';
import { GitHubClient } from '../service/github-client';
import { runWithTracing } from '../utils/tracing';

function createCacheKey(repository: Repository) {
  return `contributors-json-cache/v1.2/${repository.owner}--${repository.repo}.json`;
}

@injectable()
export class ContributorsRepository {
  constructor(private readonly githubClient: GitHubClient, private readonly cacheStorage: CacheStorage) {}

  async getContributors(repository: Repository): Promise<RepositoryContributors> {
    return runWithTracing('ContributorsRepository.getContributors', async () => {
      const cacheKey = createCacheKey(repository);
      const cached = await this.cacheStorage.restoreJSON<RepositoryContributors>(cacheKey);
      if (cached) {
        return cached;
      }

      const contributors = await this.githubClient.getContributors(repository);
      await this.cacheStorage.saveJSON(cacheKey, contributors);
      return contributors;
    });
  }
}

File: apps/api-go/service/contributors.go
package service

import (
	"context"
	"fmt"

	"contrib.rocks/apps/api-go/core"
	"contrib.rocks/apps/api-go/infrastructure"
	"contrib.rocks/libs/goutils/model"
)

type ContributorsService struct {
	cache *infrastructure.CacheStorage
	ghc   *GitHubService
}

func NewContributorsService(i *core.Infrastructure) *ContributorsService {
	return &ContributorsService{i.Cache, NewGitHubService(i)}
}

func (s *ContributorsService) GetContributors(ctx context.Context, r *model.Repository) (*model.RepositoryContributors, error) {
	cacheKey := createContributorsJSONCacheKey(r)
	// restore cache
	var cache *model.RepositoryContributors
	err := s.cache.GetJSON(ctx, cacheKey, cache)
	if err != nil {
		return nil, err
	}
	if cache != nil {
		return cache, nil
	}
	data, err := s.ghc.GetContributors(ctx, r)
	if err != nil {
		return nil, err
	}
	// save cache
	err = s.cache.SaveJSON(ctx, cacheKey, data)
	if err != nil {
		return nil, err
	}
	return data, nil
}

func createContributorsJSONCacheKey(r *model.Repository) string {
	// `contributors-json-cache/v1.2/${repository.owner}--${repository.repo}.json`;
	return fmt.Sprintf("contributors-json-cache/v1.2/%s--%s.json", r.Owner, r.RepoName)
}
